File: occonfig-node/__tests__/env.test.ts
import { initByEnvMap } from "../src/env";

describe("init by env", () => {
    type TestCase = {
        name: string
        envs: { [key: string]: string }
        serviceName?: string
        serviceUrl?: string
        zPage?: string
        configFile?: string
        traceExporter?: string
        traceSampler?: number
        statsExporter?: string
    };

    const testCases: TestCase[] = [
        {
            name: "service-name test",
            envs: {OC_SERVICE_NAME: "my-service", HOME: "test"},
            serviceName: "my-service",
        },
        {
            name: "service-url test",
            envs: {OC_SERVICE_URL: "http://localhost:8080", HOME: "test"},
            serviceUrl: "http://localhost:8080",
        },
        {
            name: "zpage test",
            envs: {OC_ZPAGE: "http://:8888/debug", HOME: "test"},
            zPage: "http://:8888/debug",
        },
        {
            name: "config-json test",
            envs: {OC_CONFIG_JSON: "config.json", HOME: "test"},
            configFile: "config.json",
        },
        {
            name: "trace-exporter test",
            envs: {OC_TRACE_EXPORTER: "jaeger://localhost:6831", HOME: "test"},
            traceExporter: "jaeger://localhost:6831",
        },
        {
            name: "trace-sampler test (1)",
            envs: {OC_TRACE_SAMPLER: "never", HOME: "test"},
            traceSampler: 0.0,
        },
        {
            name: "trace-sampler test (2)",
            envs: {OC_TRACE_SAMPLER: "always", HOME: "test"},
            traceSampler: 1.0,
        },
        {
            name: "trace-sampler test (3)",
            envs: {OC_TRACE_SAMPLER: "0.25", HOME: "test"},
            traceSampler: 0.25,
        },
        {
            name: "stats-exporter test",
            envs: {OC_STATS_EXPORTER: "prometheus://:8888", HOME: "test"},
            statsExporter: "prometheus://:8888",
        },
    ];

    testCases.forEach((testcase) => {
        it(testcase.name, () => {
            const result = initByEnvMap(testcase.envs)
            if (testcase.serviceName) {
                expect(result.serviceName).toEqual(testcase.serviceName);
            }
            if (testcase.serviceUrl) {
                expect(result.serviceUrl).toEqual(testcase.serviceUrl);
            }
            if (testcase.zPage) {
                expect(result.zPage).toEqual(testcase.zPage);
            }
            if (testcase.configFile) {
                expect(result.configFile).toEqual(testcase.configFile);
            }
            if (testcase.traceExporter) {
                expect(result.traceExporter).toEqual(testcase.traceExporter);
            }
            if (testcase.traceSampler) {
                expect(result.traceSampler).toBeCloseTo(testcase.traceSampler);
            }
            if (testcase.statsExporter) {
                expect(result.statsExporter).toEqual(testcase.statsExporter);
            }
        });
    });
});

File: occonfig/env_test.go
package occonfig

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestEnv(t *testing.T) {
	testcases := []struct {
		Name          string
		Envs          []string
		ServiceName   string
		ServiceUrl    string
		ZPage         string
		ConfigFile    string
		HoneycombKey  string
		TraceExporter string
		TraceSampler  float64
		StatsExporter string
	}{
		{
			Name:         "service-name test",
			Envs:         []string{"OC_SERVICE_NAME=my-service", "HOME=test"},
			ServiceName:  "my-service",
			TraceSampler: -1,
		},
		{
			Name:         "service-url test",
			Envs:         []string{"OC_SERVICE_URL=http://localhost:8080", "HOME=test"},
			ServiceUrl:   "http://localhost:8080",
			TraceSampler: -1,
		},
		{
			Name:         "zpage test",
			Envs:         []string{"OC_ZPAGE=http://:8888/debug", "HOME=test"},
			ZPage:        "http://:8888/debug",
			TraceSampler: -1,
		},
		{
			Name:         "config-json test",
			Envs:         []string{"OC_CONFIG_JSON=config.json", "HOME=test"},
			ConfigFile:   "config.json",
			TraceSampler: -1,
		},
		{
			Name:         "honeycomb-write-key test",
			Envs:         []string{"OC_HONEYCOMB_WRITE_KEY=honeycomb.key", "HOME=test"},
			HoneycombKey: "honeycomb.key",
			TraceSampler: -1,
		},
		{
			Name:          "trace-exporter test",
			Envs:          []string{"OC_TRACE_EXPORTER=jaeger://localhost:6831", "HOME=test"},
			TraceExporter: "jaeger://localhost:6831",
			TraceSampler:  -1,
		},
		{
			Name:         "trace-sampler test (1)",
			Envs:         []string{"OC_TRACE_SAMPLER=never", "HOME=test"},
			TraceSampler: 0.0,
		},
		{
			Name:         "trace-sampler test (2)",
			Envs:         []string{"OC_TRACE_SAMPLER=always", "HOME=test"},
			TraceSampler: 1.0,
		},
		{
			Name:         "trace-sampler test (3)",
			Envs:         []string{"OC_TRACE_SAMPLER=0.25", "HOME=test"},
			TraceSampler: 0.25,
		},
		{
			Name:          "stats-exporter test",
			Envs:          []string{"OC_STATS_EXPORTER=prometheus://:8888", "HOME=test"},
			StatsExporter: "prometheus://:8888",
			TraceSampler:  -1,
		},
	}
	for _, testcase := range testcases {
		t.Run(testcase.Name, func(t *testing.T) {
			result, err := initByEnvMap(testcase.Envs)
			assert.Nil(t, err)
			if err != nil {
				return
			}
			assert.Equal(t, testcase.ServiceName, result.ServiceName)
			assert.Equal(t, testcase.ServiceUrl, result.ServiceUrl)
			assert.Equal(t, testcase.ZPage, result.ZPage)
			assert.Equal(t, testcase.ConfigFile, result.ConfigFile)
			assert.Equal(t, testcase.HoneycombKey, result.HoneycombKey)
			assert.Equal(t, testcase.TraceExporter, result.TraceExporter)
			assert.Equal(t, testcase.StatsExporter, result.StatsExporter)
			assert.InDelta(t, testcase.TraceSampler, result.TraceSampler, 0.01)
		})
	}
}
